File: packages/lib-shared/ios/PowerAuth/PowerAuthEncryptorModule.m
/*
 * Copyright 2023 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "PowerAuthEncryptorModule.h"
#import "PowerAuthObjectRegister.h"
#import "Constants.h"
#import "Utilities.h"
#import "PAJS.h"

@import PowerAuth2;

@implementation PowerAuthEncryptorModule
{
    PowerAuthObjectRegister * _objectRegister;
}

PAJS_MODULE_REGISTRY

RCT_EXPORT_MODULE(PowerAuthEncryptor);

- (void) PAJS_INITIALIZE_METHOD
{
    PAJS_OBJECT_REGISTER
}

+ (BOOL) requiresMainQueueSetup
{
    return NO;
}

PAJS_METHOD_START(initialize,
                  PAJS_ARGUMENT(scope, NSString*)
                  PAJS_ARGUMENT(ownerId, NSString*))
{
    NSString * scopeType = [RCTConvert NSString:scope];
    BOOL activationScope;
    if ([scopeType isEqualToString:@"APPLICATION"]) {
        activationScope = NO;
    } else if ([scopeType isEqualToString:@"ACTIVATION"]) {
        activationScope = YES;
    } else {
        reject(EC_WRONG_PARAMETER, @"scope parameter is missing or contains invalid value", nil);
        return;
    }

    PowerAuthSDK * sdk = GetPowerAuthSdk(ownerId, _objectRegister, reject);
    if (!sdk) {
        return;
    }

    void (^callback)(PowerAuthEncryptor*, NSError*) = ^(PowerAuthEncryptor * encryptor, NSError * error) {
        if (!encryptor) {
            ProcessError(error, reject);
            return;
        }
        NSString * objectId = [self->_objectRegister registerObject:encryptor
                                                     ifOwnerMatches:sdk
                                                            ownerId:ownerId
                                                           policies:@[ RP_KEEP_ALIVE(ENCRYPTOR_KEEP_ALIVE_TIME) ]];
        if (!objectId) {
            reject(EC_INSTANCE_NOT_CONFIGURED, @"PowerAuth instance is no longer configured", nil);
            return;
        }
        resolve(objectId);
    };

    if (activationScope) {
        [sdk encryptorForActivationScopeWithCallback:callback];
    } else {
        [sdk encryptorForApplicationScopeWithCallback:callback];
    }
}
PAJS_METHOD_END

PAJS_METHOD_START(canEncryptRequest,
                  PAJS_ARGUMENT(encryptorId, NSString*))
{
    PowerAuthEncryptor * encryptor = [_objectRegister touchObjectWithId:encryptorId
                                                         expectedClass:[PowerAuthEncryptor class]];
    if (!encryptor) {
        reject(EC_INVALID_NATIVE_OBJECT, @"Encryptor object is no longer valid", nil);
        return;
    }
    resolve(@(encryptor.canEncryptRequest));
}
PAJS_METHOD_END

PAJS_METHOD_START(canDecryptResponse,
                  PAJS_ARGUMENT(encryptorId, NSString*))
{
    PowerAuthEncryptor * encryptor = [_objectRegister touchObjectWithId:encryptorId
                                                         expectedClass:[PowerAuthEncryptor class]];
    if (!encryptor) {
        reject(EC_INVALID_NATIVE_OBJECT, @"Encryptor object is no longer valid", nil);
        return;
    }
    resolve(@(encryptor.canDecryptResponse));
}
PAJS_METHOD_END

PAJS_METHOD_START(encryptRequest,
                  PAJS_ARGUMENT(encryptorId, NSString*)
                  PAJS_ARGUMENT(requestBodyBase64, NSString*))
{
    NSData * clearBody = requestBodyBase64
        ? [[NSData alloc] initWithBase64EncodedString:requestBodyBase64 options:0]
        : nil;
    if (requestBodyBase64 && !clearBody) {
        reject(EC_WRONG_PARAMETER, @"Request body is not valid Base64", nil);
        return;
    }
    PowerAuthEncryptor * encryptor = [_objectRegister touchObjectWithId:encryptorId
                                                         expectedClass:[PowerAuthEncryptor class]];
    if (!encryptor) {
        reject(EC_INVALID_NATIVE_OBJECT, @"Encryptor object is no longer valid", nil);
        return;
    }
    NSError * encryptionError = nil;
    PowerAuthEncryptedRequest * encryptedRequest = [encryptor encryptRequest:clearBody error:&encryptionError];
    NSString * serializedBody = encryptedRequest
        ? [encryptedRequest.requestBody base64EncodedStringWithOptions:0]
        : nil;
    if (!serializedBody) {
        ProcessError(encryptionError, reject);
        return;
    }
    NSMutableArray * headers = [NSMutableArray arrayWithCapacity:encryptedRequest.requestHeaders.count];
    for (PowerAuthHttpHeader * header in encryptedRequest.requestHeaders) {
        [headers addObject:@{ @"name": header.key, @"value": header.value }];
    }
    resolve(@{ @"requestBody": serializedBody, @"requestHeaders": headers });
}
PAJS_METHOD_END

PAJS_METHOD_START(decryptResponse,
                  PAJS_ARGUMENT(encryptorId, NSString*)
                  PAJS_ARGUMENT(responseBodyBase64, NSString*))
{
    PowerAuthEncryptor * encryptor = [_objectRegister touchObjectWithId:encryptorId
                                                         expectedClass:[PowerAuthEncryptor class]];
    NSData * bodyData = [[NSData alloc] initWithBase64EncodedString:responseBodyBase64 options:0];
    if (!bodyData) {
        [_objectRegister releaseObjectWithId:encryptorId];
        reject(EC_WRONG_PARAMETER, @"Response body is not valid Base64", nil);
        return;
    }
    if (!encryptor) {
        [_objectRegister releaseObjectWithId:encryptorId];
        reject(EC_INVALID_NATIVE_OBJECT, @"Encryptor object is no longer valid", nil);
        return;
    }
    NSError * decryptionError = nil;
    PowerAuthEncryptedResponse * encryptedResponse = [[PowerAuthEncryptedResponse alloc] initWithResponseBody:bodyData error:&decryptionError];
    NSData * clearResponse = encryptedResponse
        ? [encryptor decryptResponse:encryptedResponse error:&decryptionError]
        : nil;
    NSString * result = [clearResponse base64EncodedStringWithOptions:0];

    [_objectRegister releaseObjectWithId:encryptorId];

    if (!result) {
        ProcessError(decryptionError, reject);
        return;
    }
    resolve(result);
}
PAJS_METHOD_END

@end

File: packages/lib-rn/android/src/main/java/com/wultra/android/powerauth/reactnative/PowerAuthEncryptorModule.java
/*
 * Copyright 2023 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.wultra.android.powerauth.reactnative;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.BaseJavaModule;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.module.annotations.ReactModule;
import com.wultra.android.powerauth.js.PowerAuthEncryptorJsModule;

@SuppressWarnings("unused")
@ReactModule(name = "PowerAuthEncryptor")
public class PowerAuthEncryptorModule extends BaseJavaModule {

    private final ReactApplicationContext context;
    private final ObjectRegister objectRegister;
    private final PowerAuthEncryptorJsModule powerAuthEncryptorJsModule;

    public PowerAuthEncryptorModule(@NonNull ReactApplicationContext context, @NonNull ObjectRegister objectRegister) {
        super();
        this.context = context;
        this.objectRegister = objectRegister;
        this.powerAuthEncryptorJsModule = new PowerAuthEncryptorJsModule(context, objectRegister.getObjectRegisterJs());
    }

    public PowerAuthEncryptorJsModule getPowerAuthEncryptorJsModule() {
        return powerAuthEncryptorJsModule;
    }

    @NonNull
    @Override
    public String getName() {
        return powerAuthEncryptorJsModule.getName();
    }

    public static final String NAME = "PowerAuthEncryptor";

    // JavaScript methods

    @ReactMethod
    void initialize(@NonNull String scope, @NonNull String ownerId, Promise promise) {
        powerAuthEncryptorJsModule.initialize(scope, ownerId, promise);
    }

    // Encryption

    @ReactMethod
    void canEncryptRequest(@NonNull String encryptorId, Promise promise) {
        powerAuthEncryptorJsModule.canEncryptRequest(encryptorId, promise);
    }

    @ReactMethod
    void encryptRequest(@NonNull String encryptorId, @Nullable String requestBodyBase64, Promise promise) {
        powerAuthEncryptorJsModule.encryptRequest(encryptorId, requestBodyBase64, promise);
    }

    // Decryption

    @ReactMethod
    void canDecryptResponse(String encryptorId, Promise promise) {
        powerAuthEncryptorJsModule.canDecryptResponse(encryptorId, promise);
    }

    @ReactMethod
    void decryptResponse(@NonNull String encryptorId, @NonNull String responseBody, Promise promise) {
        powerAuthEncryptorJsModule.decryptResponse(encryptorId, responseBody, promise);
    }
}
